src/cart/src/cart/crt_rpc.c

```diff
@@ -536,224 +536,222 @@ crt_req_fill_tgt_uri(struct crt_rpc_priv *rpc_priv, crt_phy_addr_t base_uri)
+retry:
+	/* TODO: add retry logic for CART-688 */
+
 out:
+	RPC_PUB_DECREF(lookup_rpc);
+
+	/* Force complete and destroy chained rpc */
 	if (rc != 0) {
-		crt_context_req_untrack(rpc_priv);
-		crt_rpc_complete(rpc_priv, rc);
-		RPC_DECREF(rpc_priv); /* destroy */
+		crt_context_req_untrack(chained_rpc_priv);
+		crt_rpc_complete(chained_rpc_priv, rc);
+		RPC_DECREF(chained_rpc_priv);
 	}
 
-	if (!ul_retried && rpc_priv->crp_ul_req != NULL) {
-		/* addref in crt_req_uri_lookup_by_rpc */
-		RPC_PUB_DECREF(rpc_priv->crp_ul_req);
-		rpc_priv->crp_ul_req = NULL;
+	/* Addref done in crt_issue_uri_lookup() */
+	RPC_DECREF(chained_rpc_priv);
+}
+
+/**
+ * Helper function that returns rank for clients to contact
+ * for performing uri lookups
+ */
+static d_rank_t
+crt_client_get_contact_rank(crt_context_t crt_ctx, crt_group_t *grp,
+			d_rank_t query_rank, uint32_t query_tag)
+{
+	struct crt_grp_priv	*grp_priv;
+	d_rank_t		contact_rank = CRT_NO_RANK;
+	char			*cached_uri = NULL;
+	struct crt_context	*ctx;
+	int			rc;
+
+	grp_priv = crt_grp_pub2priv(grp);
+	ctx = crt_ctx;
+
+	/* If query_rank:tag=0 is in cache, use it as contact destination */
+	if (query_tag != 0) {
+		rc = crt_grp_lc_lookup(grp_priv, ctx->cc_idx,
+				query_rank, 0, &cached_uri, NULL);
+		if (rc == 0 && cached_uri != NULL)
+			D_GOTO(out, contact_rank = query_rank);
 	}
-	/* addref in crt_req_uri_lookup_by_rpc */
-	RPC_DECREF(rpc_priv);
+
+	D_RWLOCK_RDLOCK(&grp_priv->gp_rwlock);
+	/* TODO: Add logic for CART-688 */
+	contact_rank = grp_priv->gp_psr_rank;
+	D_RWLOCK_UNLOCK(&grp_priv->gp_rwlock);
+
+out:
+	return contact_rank;
 }
 
 static int
-crt_req_uri_lookup_by_rpc(struct crt_rpc_priv *rpc_priv, crt_cb_t complete_cb,
-			  void *arg, d_rank_t rank, uint32_t tag)
+crt_req_uri_lookup(struct crt_rpc_priv *rpc_priv)
 {
-	crt_rpc_t			*ul_req;
-	crt_endpoint_t			*orig_tgt_ep;
-	crt_endpoint_t			 ul_tgt_ep = {0};
-	struct crt_grp_priv		*grp_priv;
-	struct crt_uri_lookup_in	*ul_in;
-	struct crt_uri_lookup_out	*ul_out;
-	int				 rc = 0;
-
-	orig_tgt_ep = &rpc_priv->crp_pub.cr_ep;
+	crt_endpoint_t	*tgt_ep;
+	crt_context_t	ctx;
+	crt_group_t	*grp;
+	int		rc;
 
-	grp_priv = crt_grp_pub2priv(orig_tgt_ep->ep_grp);
+	tgt_ep = &rpc_priv->crp_pub.cr_ep;
+	ctx = rpc_priv->crp_pub.cr_ctx;
+	grp = tgt_ep->ep_grp;
 
-	ul_tgt_ep.ep_grp = orig_tgt_ep->ep_grp;
-	ul_tgt_ep.ep_rank = rank;
-	ul_tgt_ep.ep_tag = tag;
+	/* Client handling */
+	if (!crt_is_service()) {
+		d_rank_t lookup_rank;
+
+		lookup_rank = crt_client_get_contact_rank(ctx, grp,
+						tgt_ep->ep_rank,
+						tgt_ep->ep_tag);
+		if (lookup_rank == CRT_NO_RANK) {
+			D_ERROR("Failed to rank for uri lookups\n");
+			D_GOTO(out, rc = -DER_NONEXIST);
+		}
 
-	rc = crt_req_create(rpc_priv->crp_pub.cr_ctx, &ul_tgt_ep,
-			    CRT_OPC_URI_LOOKUP, &ul_req);
-	if (rc != 0) {
-		D_ERROR("crt_req_create URI_LOOKUP failed, rc: %d opc: %#x.\n",
-			rc, rpc_priv->crp_pub.cr_opc);
+		rc = crt_issue_uri_lookup(ctx, grp,
+					lookup_rank, 0,
```

Comment:
```suggestion
					  lookup_rank, 0,
```

Reply:
These look like they'd fit on the same line.